description/src/transformer.rs

```diff
@@ -0,0 +1,72 @@
+use std::{cell::RefCell, collections::HashMap};
+
+use scale_info::{form::PortableForm, PortableRegistry, Type};
+
+/// The transformer provides an abstraction for traversing a type registry
+/// given a type_id as a starting point, and **transforming** it into a tree-like structure.
+/// It provides a cache that shields users from infinite recursion.
+///
+/// In this way, we can have easy recursion protection mechanisms for type descirptions, rust type examples and scale value type examples.
+pub struct Transformer<'a, R, S = ()> {
+    /// keep this private such that the cache is sealed and connot be accessed from outside of the Transformer::transform function
+    cache: RefCell<HashMap<u32, Cached<R>>>,
+    /// state can be used for example for an Rng
+    pub state: S,
```

Comment:
Can this be made private? 

You have already added a `pub state` which should be sufficient?